### DISCUSSION TITLE
Not able to load mantine css inside iframe

### DISCUSSION THREAD

**Discussion by 318097** [mantinedev/mantine#598] [Q&A] [answered]
I really love the @mantine/core package. I have built the entire side project and now have to load that in an iframe. 
I achieved this using `react-frame-component` and the JS loads fine, inserted the local project's CSS in the iframe using `mini-css-extract-plugin` but not able to get mantine core components CSS. 
Without Iframe also, the CSS seems to be imported from `<style data-jss="" data-meta="Unthemed"></style>`.

How to get the hosted CSS or where in DOM is the stylesheet for mantine packages?

**Comment by rtivital** [comment]
Mantine does not use JSS anymore, please update to the latest version and see if that resolves your issue.

**Reply by 318097** [reply]
@rtivital - Cool will try. But is there a hosted CSS URL? Also can you explain more about `<style data-jss="" data-meta="Unthemed"></style>`. I don't understand what this does or mean.

**Reply by rtivital** [reply]
Mantine uses css-in-js, all styles are included in js bundle and added to head via style tag in runtime

**Reply by 318097** [reply]
So that must be the reason for CSS not working. Cool, I'll try with the new version. Thanks :)

**Reply by 318097** [reply]
@rtivital - I have updated all my packages to `3.4.3` now the styles are loaded in the main DOMs head. It's not loading in the iframe's head. Any other method you can recommend?

**Reply by rtivital** [reply]
No, idea, maybe Mantine is just incompatible with that library. I would start by exploring react-frame-component docs and asking maintainer of that library for help.

**Reply by 318097** [reply]
Ok thanks

**Reply by 318097** [reply]
Is the CSS hosted on a CDN?

**Reply by rtivital** [reply]
No, all styles are included in js bundle and added to head via style tag in runtime

**Reply by ryanseddon** [reply]
So what mantine needs to do is expose the underlying [`createCache`](https://emotion.sh/docs/@emotion/cache) from emotion so some configuration can be changed including where the styles are injected e.g. you do expose the `getCache` but that is just a readonly call.

```js
import createCache from '@emotion/cache'
import { CacheProvider } from '@emotion/core'

import Frame, { FrameContextConsumer } from 'react-frame-component'

export default ({ children }) => {
  return (
    <Frame>
      <FrameContextConsumer>
        {{document} => {
          // Configure emotion to target iframes head using the frame context
          const cache = createCache({
            container: document.head
          });

          return (
            <CacheProvider value={cache}>
              {children}
            </CacheProvider>
          )
        }}
      </FrameContextConsumer>
    </Frame>
  )
}
```

Essentially the `CacheProvider` and `createCache` will need to be exposed via mantine. Doing this also allows you to support CSP when `nonce` attributes need to be passed down to injected styles.

**Comment by porkopek** [comment, marked as answer]
My case is not an iframe, but the shadow-dom, but I think is the same.
What worked for me was
```jsx

//creates the shadow dom
const element = document.createElement('div');
const shadowRoot = element.attachShadow({ mode: 'open' });
const mountPoint = document.createElement('div');
const emotionRoot = document.createElement('div');

shadowRoot.appendChild(mountPoint);
shadowRoot.appendChild(emotionRoot);

document.body.appendChild(element);


ReactDOM.render(
  <MantineProvider
    defaultProps={{ Portal: { target: emotionRoot } }} // this is necessary, because Portal renders by default outside of shadow-dom
    emotionOptions={{
      container: emotionRoot, // with this line, emotion renders styles inside the shadow dom element created before
      key: 'mantine',
    }}
  >
   Hello World!
  </MantineProvider>,
  mountPoint
);
```

**Reply by rtivital** [reply]
Thanks for sharing!

**Reply by kyawkhaing-genie** [reply]
defaultProps and emotionOptions are not existed in mantine/core version 5.3.0. Can you help me explain how to inject mantine styles to shadow dom?

**Reply by rtivital** [reply]
- https://mantine.dev/theming/emotion-cache/
- https://mantine.dev/theming/default-props/

**Reply by kyawkhaing-genie** [reply]
Thank you! It works well.

**Reply by hyusetiawan** [reply]
apologies for bringing up old thread but this answers exactly the issue im facing, I am not familiar with the recent changes on the API but how do I do this for the latest mantine 7.4.1?
here are my attempts: 
the default props moved to theme, so:
```

export const theme = createTheme({
  components: {
    Modal: {
      defaultProps: {
        target: emotionRoot,
      },
    },
  },
})
```
what to do with `emotionOptions`?

**Reply by hyusetiawan** [reply]
ah turns out mantine 7 does not use emotion anymore, is there a way to do it for version 7?
cc @rtivital @porkopek

**Reply by rtivital** [reply]
Starting from version 7, Mantine provides a CSS file that must be included in your application. It can be included any way your application requires, either imported with a bundler or with a link HTML tag.

**Reply by hyusetiawan** [reply]
this question is for sure outside of the scope of Mantine, but if you dont't mind, would it be possible to share what that looks like?
My attempt is below:
```
import { createRoot } from "react-dom/client"
import css from "@mantine/core/styles.css?inline"
import {
  Button,
  MantineProvider,
  createTheme,
} from "@mantine/core"
import { randFullAddress } from "@ngneat/falso"
export const theme = createTheme({
  /** Put your mantine theme override here */
})

const hostElement = document.createElement("div")

// Attach a shadow root to the element
const shadowRoot = hostElement.attachShadow({ mode: "closed" })

const appElement = document.createElement("div")

shadowRoot.appendChild(appElement)

document.body.appendChild(hostElement)

const root = createRoot(appElement)
root.render(
  <>
    <style dangerouslySetInnerHTML={{ __html: css }} />
    <MantineProvider getRootElement={() => appElement} theme={theme}>
      <div
        style={{
          position: "fixed",
          bottom: 20,
          left: 20,
          right: 20,
          zIndex: 30000,
        }}
      >
        <Button variant="outline" color="blue">
          hello
        </Button>
      </div>
    </MantineProvider>
  </>
)

```

I am assuming that It is missing the CSS vars, ex: `var(--mantine-color-blue-outline)` i am sure i am just missing a trivial step somewhere

**Reply by hyusetiawan** [reply]
another attempt, which is SUPER HACKY but somehow gets me closer:
```
const processedStyle = css
  .replaceAll(":root", ":host")
```
it seems the selector :root should be host as it refers to the shadow root? my shadow dom knowledge is lacking here, but 
`:root[data-mantine-color-scheme="light"]` selector, or should I say `:host[data-mantine-color-scheme="light"]` is not getting triggered probably due to data-mantine-color-scheme is not set?

**Reply by rtivital** [reply]
1. https://mantine.dev/theming/mantine-provider/#cssvariablesselector
2. https://mantine.dev/theming/mantine-provider/#getrootelement

**Reply by hyusetiawan** [reply]
@rtivital thank you, cssvariableselector is the missing piece, for people looking a setup using shadow dom with mantine, here is a working boilerplate as of mantine 7.
```
import { createRoot } from "react-dom/client"
import css from "@mantine/core/styles.css?inline"
import { Button, MantineProvider, createTheme } from "@mantine/core"
import ContentApp from "@/ContentApp"
export const theme = createTheme({
  /** Put your mantine theme override here */
})

const hostElement = document.createElement("div")

// Attach a shadow root to the element
const shadowRoot = hostElement.attachShadow({ mode: "closed" })

const appElement = document.createElement("div")
appElement.setAttribute("id", "app")
shadowRoot.appendChild(appElement)
// document.body.appendChild(appElement)
document.body.appendChild(hostElement)
const root = createRoot(appElement)

console.log(css)
root.render(
  <body>
    <MantineProvider
      getRootElement={() => appElement}
      cssVariablesSelector="#app"
      //   withCssVariables={false}
      theme={theme}
    >
      <style dangerouslySetInnerHTML={{ __html: css }} />
      <Button>works</Button>
    </MantineProvider>
  </body>
)
```

**Reply by hjhjdev** [reply]
how about css modules? using vite with any "~.module.css" doesn't work.
Just adding classes.classname strategy doesn't work.

**Reply by laugri** [reply]
For anyone interested in making Mantine work with content-script UIs in browser extensions, with the wrt.dev framework, here is what my `entrypoints/context/index.tsx` code looks like:

```tsx
import "@mantine/core/styles.css"
import './style.css';
import ReactDOM from 'react-dom/client';
import { MantineProvider } from '@mantine/core';
import App from './App';

export default defineContentScript({
  matches: ['*://*/*'],
  cssInjectionMode: 'ui',

  async main(ctx) {
    const ui = await createShadowRootUi(ctx, {
      name: 'bloom-care-plugin',
      position: 'overlay',
      anchor: 'body',
      append: 'first',
      onMount: (container) => {
        // Container is a body, and React warns when creating a root on the body, so create a wrapper div
        const app = document.createElement('div');
        // in order to give Mantine a cssVariablesSelector, add an id or some kind of selector to the wrapper
        app.setAttribute("id", "bloom-care-extension-root")
        container.append(app);
        // Create a root on the UI container and render a component
        const root = ReactDOM.createRoot(app);
        root.render(
          <>
            <MantineProvider
              // set cssVariablesSelector and getRootElement to refer to the wrapper
              cssVariablesSelector="#bloom-care-extension-root"
              getRootElement={() => app}
            >
              <App />
            </MantineProvider>
          </>
        );
        return root;
      },
      onRemove: (root) => {
        root?.unmount();
      },
    });

    ui.mount();
  },
});
```

Many thanks to @hyusetiawan !

**Reply by laugri** [reply]
EDIT:

To make Modal, DatepickerInput, Select, and other Portal-dependant components: always use the `withinPortal: false` property (e.g. ([for popover](https://mantine.dev/core/popover/#nested-popovers))) since portals don't work with with ShadowDom. 

I couldn't make the Portal target work with ShadowDom (portals were either outside of it, or couldn't find selectors inside of it).

---

Original comment:

@rtivital Hi. Isn't Portal extendable in Mantine v7 ? I tried porting @porkopek 's suggestion of updated the target for portals but getting an error  `Property 'extend' does not exist on type 'ForwardRefExoticComponent<PortalProps & RefAttributes<HTMLDivElement>>`. Am I doing something incorrectly ?

What I am trying to adap to v7 from Porkopek's code:
```
defaultProps={{ Portal: { target: emotionRoot } }} // this is necessary, because Portal renders by default outside of shadow-dom
```

What I am doing:
```tsx
      const customTheme = createTheme({
          components: {
            Portal: Portal.extend({ // Property 'extend' does not exist on type 'ForwardRefExoticComponent<PortalProps & RefAttributes<HTMLDivElement>>'.
              defaultProps: {
                target: root
              },
            }),
          },
        });
 ```

**Reply by Onvix24** [reply]
Thanks you.

**Comment by Kinbaum** [comment]
@rtivital Looking for some guidance along this topic. I am federating my application into a shadow DOM container, and I am having some issue with the DatePicker. When I click on the input field, the DatePicker opens just fine, but when I try clicking a date or any of the controls, the picker immediately closes without making a selection.

Here is my configuration:

```jsx
// `props.root` is the shadow DOM element
const cache = useMemo(() => createEmotionCache({
  key: 'orion',
  container: props.root
}), []);

return (
  <MantineProvider
      emotionCache={ cache }
      theme={ {
          components: {
              Portal: {
                  defaultProps: {
                      target: props.root
                  }
              }
          }
      } }
      withCSSVariables
      withGlobalStyles
      withNormalizeCSS>
      <DatePicker
          label="Test"
          inputFormat="MM/DD/YYYY"
          labelFormat="MM/YYYY"
          my="lg"
          clearable={ false }
          withAsterisk
          withinPortal />
  </MantineProvider>
);
```

I am able to use my keyboard to make selections... did notice this error after pressing the tab key:

```    
Uncaught TypeError: Cannot read properties of undefined (reading 'display')
  at hidden (tabbable.js:7:1)
  at visible (tabbable.js:19:1)
  at focusable (tabbable.js:37:1)
  at tabbable (tabbable.js:42:1)
  at Array.filter (<anonymous>)
  at findTabbableDescendants (tabbable.js:45:1)
  at scopeTab (scope-tab.js:4:43)
  at HTMLDocument.handleKeyDown (use-focus-trap.js:53:17)
```

Here is a recording of this behavior

https://user-images.githubusercontent.com/6590356/202222812-722cfb98-a225-48de-b084-ed1c90184b4f.mov


### Update

I have also tried this without using the `withinPortal` property, and I still get the same type of behavior

**Reply by rtivital** [reply]
The error in console was fixed in one of the previous releases, make sure that you use latest version. Popover is clicked because it incorrectly detects outside clicks, I do not know how to fix that for shadow dom.

**Reply by Kinbaum** [reply]
Thanks for the reply. I am on the lastest version of all packages (5.8.0) and that error is still happening

**Reply by Kinbaum** [reply]
@rtivital I've managed to locate where the issue is taking place. It is in the `use-click-outside` hook:

https://github.com/mantinedev/mantine/blob/3fd5ae81f96d7f471839fbc2fa518b61228ab1e4/src/mantine-hooks/src/use-click-outside/use-click-outside.ts#L18

Rather than using `!node.contains(target))`, we can accommodate the shadow DOM by changing that line to the following:

```js
const shouldTrigger = nodes.every((node) => !!node && !event.composedPath().includes(node));
```

I'm happy to create a PR for this. It is working just as I would expect, and the original behavior is preserved as well.

Still have to resolve the broken focus trap, and the aria hider adding the aria-hidden attribute to the wrong element

**Reply by rtivital** [reply]
Sure, feel free to send a PR

**Reply by Kinbaum** [reply]
Alright, please have a look at the changes proposed #2982

**Reply by hyusetiawan** [reply]
@Kinbaum  apologies for bringing up an old thread but i am facing this issue right now, although your fix is merged, it's somehow still exhibiting the same error.
The checkbox event gets triggered but it also closes the Popover, for completeness sake I also tried it with `withinPortal={false}` but it's still not working. Do you have leads on how to fix or work around this issue?

https://github.com/mantinedev/mantine/assets/1062920/b1a9fba0-4856-4806-95ff-7905fc16cc64
